### PR TITLE
MAINT: Run PTH flake test in prep for supporting pathlib

### PR DESCRIPTION
**Pull request by MridulS**
This PR starts work on making a unified i/o interface for reading and writing to disk. In some places using strings for filenames work, in other pathlib objects works too. We should be able to pass around `Path` objects all around the code. For a first go I want to make the `PTH` ruff flake8 linter happy.

As a start this fixes ruff complaints in couple of sub packages. Didn't want to make the PR too big.

Also want to check if this works with windows CI all fine. (`pathlib` has had some issues with windows from what I remember).

One way to go about https://github.com/astropy/astropy/issues/14893

### CODE REVIEW

File: astropy/time/setup_package.py
Comment (nstarman) [inline astropy/time/setup_package.py:11]: Relpath can be replaced.

File: astropy/time/setup_package.py
Comment (nstarman) [inline astropy/time/setup_package.py:13]: This iterates over a 1-elt list and can be simplified.

File: astropy/timeseries/periodograms/bls/setup_package.py
Comment (nstarman) [inline astropy/timeseries/periodograms/bls/setup_package.py:9]: Relpath can be replaced with relative_to

File: astropy/timeseries/periodograms/lombscargle/setup_package.py
Comment (nstarman) [inline astropy/timeseries/periodograms/lombscargle/setup_package.py:9]: Relpath can be replaced.

File: astropy/timeseries/periodograms/lombscargle/setup_package.py
Comment (nstarman) [inline astropy/timeseries/periodograms/lombscargle/setup_package.py:16]: Does Extension not work with Path objects?

File: astropy/visualization/tests/test_lupton_rgb.py
Comment (nstarman) [inline astropy/visualization/tests/test_lupton_rgb.py:186]: This test can be lightly refactored to use `tmp_path` instead of explicitly making a named temp file.

https://docs.pytest.org/en/6.2.x/tmpdir.html

File: astropy/wcs/setup_package.py
Comment (nstarman) [inline astropy/wcs/setup_package.py:17]: Relpath can be replaced.

File: astropy/wcs/tests/test_wcs.py
Comment (nstarman) [inline astropy/wcs/tests/test_wcs.py:472]: This can be 1 line

**Review by nstarman — CHANGES_REQUESTED** [review]
Lots of good things in this PR! I have a few suggestions that are applicable in many places to further simplify the code, e.g. using `read_text`.

File: astropy/timeseries/periodograms/lombscargle/setup_package.py
Comment (MridulS) [reply astropy/timeseries/periodograms/lombscargle/setup_package.py:16]: Nope, https://github.com/pypa/distutils/blob/e651e536ef1fbafbe48856815602127df0371835/distutils/extension.py#L109

I guess I can try getting this fixed upstream :)

This problem also pops up for `extension_helpers`, I'm prepping a PR to move to pathlib.

File: astropy/visualization/tests/test_lupton_rgb.py
Comment (MridulS) [reply astropy/visualization/tests/test_lupton_rgb.py:186]: I will do the refactor for test in a new PR :)

**Review by taldcroft — APPROVED** [review]
Looks good for `time`.

File: astropy/modeling/tests/test_fitters.py
Comment (eerovaher) [inline astropy/modeling/tests/test_fitters.py:239]: Does this fit in a single line?
```suggestion
        reclist = Path(get_pkg_data_filename("data/idcompspec.fits")).read_text().split("begin")
```

File: astropy/modeling/tests/test_projections.py
Comment (eerovaher) [inline astropy/modeling/tests/test_projections.py:37]: Shouldn't this be making use of `__file__`?

File: astropy/time/tests/test_update_leap_seconds.py
Comment (eerovaher) [inline astropy/time/tests/test_update_leap_seconds.py:66]:
```suggestion
        Path(bad_file).write_text("\n".join(line for line in lines if not line.startswith("#")))
```

File: astropy/time/tests/test_update_leap_seconds.py
Comment (eerovaher) [inline astropy/time/tests/test_update_leap_seconds.py:77]: Here too it looks like it would be possible to use `write_text()`.

File: astropy/visualization/scripts/fits2bitmap.py
Comment (eerovaher) [inline astropy/visualization/scripts/fits2bitmap.py:107]:
```suggestion
        out_fn = filename.with_suffix("")
```

File: astropy/visualization/scripts/fits2bitmap.py
Comment (eerovaher) [inline astropy/visualization/scripts/fits2bitmap.py:111]: If `out_fn` still has a suffix then the new code is not equivalent to the old code.

File: astropy/wcs/tests/test_wcs.py
Comment (eerovaher) [inline astropy/wcs/tests/test_wcs.py:742]:
```suggestion
    lines = testfile.read_text().splitlines()
```

File: astropy/wcs/tests/test_wcs.py
Comment (eerovaher) [inline astropy/wcs/tests/test_wcs.py:751]:
```suggestion
    lines = testfile.read_text().splitlines()
```

File: astropy/modeling/tests/test_polynomial.py
Comment (nstarman) [inline astropy/modeling/tests/test_polynomial.py:354]: I don't think the forward slash is correct on all systems. Can the function not accept a Path argument?

File: astropy/wcs/wcsapi/fitswcs.py
Comment (eerovaher) [inline astropy/wcs/wcsapi/fitswcs.py:156]: This is off-topic for this pull request.

File: astropy/wcs/wcsapi/low_level_api.py
Comment (eerovaher) [inline astropy/wcs/wcsapi/low_level_api.py:347]:
```suggestion
UCDS_FILE = Path(__file__).parent / "data" / "ucds.txt"
```

File: astropy/time/setup_package.py
Comment (nstarman) [inline astropy/time/setup_package.py:12]:
```suggestion
SRC_FILES = [str(C_TIME_PKGDIR / "src" / "parse_times.c")]
```

File: astropy/wcs/wcsapi/low_level_api.py
Comment (eerovaher) [inline astropy/wcs/wcsapi/low_level_api.py:350]:
```suggestion
VALID_UCDS = {x.strip() for x in UCDS_FILE.read_text().splitlines()[1:]}
```

File: astropy/wcs/tests/test_pickle.py
Comment (nstarman) [inline astropy/wcs/tests/test_pickle.py:93]:
```suggestion
    header = get_pkg_data_contents(Path("data")/"outside_sky.hdr"), encoding="binary")
```

File: astropy/wcs/tests/test_profiling.py
Comment (nstarman) [inline astropy/wcs/tests/test_profiling.py:69]:
```suggestion
    header = get_pkg_data_contents(Path("data")/"spectra") / filename)
```

**Review by nstarman — CHANGES_REQUESTED** [review]
Some of the `os.path.join` got replaced by os-dependent strings.

File: astropy/logger.py
Comment (neutrinoceros) [reply astropy/logger.py:224]: better yet, if we don't use the key, we could iterate directly on `.values()`

File: astropy/modeling/tests/test_polynomial.py
Comment (MridulS) [reply astropy/modeling/tests/test_polynomial.py:354]: The `get_pkg_data_*` do not work with path objects, I plan on creating a new PR to fix that. This PR I want to narrow down on making ruff happy.

> I don't think the forward slash is correct on all systems.

Yes but shouldn't the CI tests catch them? The windows CI test seem to be passing?

File: astropy/wcs/setup_package.py
Comment (mcara) [inline astropy/wcs/setup_package.py:16]: I don't see how using `Path` here improves anything. `parent`, `resolve`, `cwd` are all extra steps I need to think about when reading this line.

File: astropy/modeling/tests/test_polynomial.py
Comment (nstarman) [reply astropy/modeling/tests/test_polynomial.py:354]: > The get_pkg_data_* do not work with path objects, I plan on creating a new PR to fix that. This PR I want to narrow down on making ruff happy.

SGTM.

> Yes but shouldn't the CI tests catch them? The windows CI test seem to be passing?

Depends on exact usage. See https://docs.python.org/3/library/os.path.html#os.path.join for details.

File: astropy/time/setup_package.py
Comment (nstarman) [reply astropy/time/setup_package.py:12]: Might as well improve these path constructions while we're here.

File: astropy/wcs/setup_package.py
Comment (nstarman) [reply astropy/wcs/setup_package.py:16]: `.parent` = `os.path.dirname`
`relative_to(Path.cwd())` = `os.path.relpath` (which makes a relative path, but to what? It's given only 1 argument, so what it's relative to isn't obvious. Reading the docs it has a second argument, `start=os.curdir` which then makes this exactly the same as the more explicit `.relative_to(Path.cwd())`).

So as far as I can tell, the major difference is only `.resolve()`.

File: astropy/visualization/tests/test_lupton_rgb.py
Comment (nstarman) [reply astropy/visualization/tests/test_lupton_rgb.py:186]: SGTM

File: astropy/wcs/setup_package.py
Comment (eerovaher) [reply astropy/wcs/setup_package.py:16]: I don't think `resolve()` is even needed here because `Path(__file__)` already produces an absolute path.

File: astropy/modeling/tests/test_fitters.py
Comment (nstarman) [reply astropy/modeling/tests/test_fitters.py:239]: If not yet, then when `get_pkg_data_filename` is upgraded to return a Path...

Comment (MridulS) [reply astropy/modeling/tests/test_fitters.py:239]: I can start working on this in this PR if that makes it easier to review. My plan was to get the `get_pkg_data_*` methods work with pathlib natively in a new PR.

Comment (nstarman) [reply astropy/modeling/tests/test_fitters.py:239]: No need to do it in this PR.

Comment (nstarman) [reply astropy/modeling/tests/test_fitters.py:239]: I was just commenting that it will reduce the line length, maybe making it 1 line in the future.

File: astropy/coordinates/name_resolve.py
Comment (eerovaher) [reply astropy/coordinates/name_resolve.py:160]: @pllim, could you run the following on Windows?
```python
from astropy.coordinates import SkyCoord
SkyCoord.from_name("Fomalhaut")
```

Comment (pllim) [reply astropy/coordinates/name_resolve.py:160]: I can later, unless someone beats me to it. Surely I am not the only one using Windows? 😆

**Comment by eerovaher** [comment]
The way I understand it the intent behind the Ruff PTH rules is that if code is representing filesystem paths as `Path` instances then it's best to commit to doing that and to avoid converting them to `str` and then back to `Path` again. But what seems to be happening in this pull request is that some of the code involving our `get_pkg_data_*` functions is performing exactly this kind of back and forth shuffling between `Path` and `str` that the Ruff PTH rules are meant to guard against. It might therefore be better to ensure that the `get_pkg_data_*` functions can handle `Path` instances first, and then it should become possible for the relevant code to use `Path` throughout.

**Comment by neutrinoceros** [comment]
There might be a slight misunderstanding here, that not what I understand the intention of this rule set to be. Here's the one liner description for `flake8-use-pathlib` (which ruff reimplements as PTH):
> A plugin for flake8 finding use of functions that can be replaced by pathlib module.

File: astropy/coordinates/name_resolve.py
Comment (eerovaher) [reply astropy/coordinates/name_resolve.py:160]: @pllim, have you had the time to test this in Windows?

File: docs/conf.py
Comment (mhvk) [inline docs/conf.py:153]: Just keep `open` - or use `Path(...).read_text()` or so

File: docs/conf.py
Comment (mhvk) [inline docs/conf.py:316]: keep open or just get the lines.

**Review by mhvk — COMMENTED** [review]
Generally, this is a good idea and very nice. That said, quite a few comments. A number are just that if a plain string with no directory is given, `open` is perfectly fine and reasonable, so no need to replace.

A more general one is whether we can pass "file names" through `Path` unconditionally; certainly, any time we do, we should adjust the docstring.

Finally, I'd hope we can get rid of a lot of `str(path)`, by ensuring `get_package_data_*` work - would seem worth doing that here, given how many people have now looked at this...

**Comment by MridulS** [comment]
> Finally, I'd hope we can get rid of a lot of str(path), by ensuring get_package_data_* work - would seem worth doing that here, given how many people have now looked at this...

Sure, I am happy to get this into this PR.

Should the changes be backward compatible? I am currently working on making the util functions work with the `str` and `Path` objects. Maybe it makes sense to just have a hard break here? (or maybe not as these things are up in the docs https://docs.astropy.org/en/latest/utils/data.html)

If not a hard break, does it make sense to deprecate usage of `str` only args for future removal?

File: astropy/time/tests/test_update_leap_seconds.py
Comment (MridulS) [reply astropy/time/tests/test_update_leap_seconds.py:62]: It can't, a lot of astropy code is hardcoded to using to path strings.

File: astropy/wcs/setup_package.py
Comment (MridulS) [reply astropy/wcs/setup_package.py:146]: This needs to be fixed in extension-helpers, I have started a PR https://github.com/astropy/extension-helpers/pull/83

**Comment by mhvk** [comment]
For this PR at least, I don't think we should have anything that is not backwards compatible. But presumably passing a `path` through `Path` is super fast.

File: astropy/time/tests/test_update_leap_seconds.py
Comment (mhvk) [reply astropy/time/tests/test_update_leap_seconds.py:63]: OK, leave that one for later then!

File: astropy/nddata/tests/test_ccddata.py
Comment (eerovaher) [reply astropy/nddata/tests/test_ccddata.py]: This test is checking that the `CCDData` I/O-functions perform [tilde expansion like a POSIX shell](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01). `Path` does not do tilde expansion automatically and instead expects the user to call the [`Path.expanduser()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.expanduser) method explicitly.

It is true that the code should not be converting `filename` to a `Path` because it already is one.

File: astropy/wcs/wcs.py
Comment (mcara) [inline astropy/wcs/wcs.py:449]: why is this here if it is not needed? I guess, I do not understand the comment.

Comment (MridulS) [reply astropy/wcs/wcs.py:449]: Sorry I should have been clear in the comment, in an ideal world, we should never really need to do cast Path objects to strings but a lot of code in astropy was written to work with strings.

It's becoming a big PR and I was worried about changing too many things at once. I plan on either fixing it in this PR or leave a much more informative FIXME for the next one :)

File: docs/conftest.py
Comment (mcara) [inline docs/conftest.py:38]: is `os.getcwd()` getting deprecated?

Comment (MridulS) [reply docs/conftest.py:38]: No, this is to make to standardized usage of PathLike objects rather than `str`.

(It's fine either way, but to not litter `noqa` all around we could make this change.

File: astropy/config/tests/test_configs.py
Comment (nstarman) [inline astropy/config/tests/test_configs.py:473]: just to confirm, the plan is to have a followup PR that means we don't need to do str casting?

File: astropy/convolution/setup_package.py
Comment (nstarman) [inline astropy/convolution/setup_package.py:21]: It would be nice if setuptools supported Paths for path-like things. Lot of conversations in this PR, did you submit a PR to setuptools? Either way, I'm fine to with this.

File: astropy/coordinates/tests/accuracy/generate_ref_ast.py
Comment (nstarman) [inline astropy/coordinates/tests/accuracy/generate_ref_ast.py:64]:
```suggestion
    with Path("data" / fnout).open("wb") as f:
```

File: astropy/coordinates/tests/accuracy/generate_ref_ast.py
Comment (nstarman) [inline astropy/coordinates/tests/accuracy/generate_ref_ast.py:130]:
```suggestion
    with Path("data" / fnout).open("wb") as f:
```

File: astropy/coordinates/tests/accuracy/generate_ref_ast.py
Comment (nstarman) [inline astropy/coordinates/tests/accuracy/generate_ref_ast.py:192]:
```suggestion
    with Path("data" / fnout)open("wb") as f:
```

File: astropy/coordinates/tests/accuracy/generate_ref_ast.py
Comment (nstarman) [inline astropy/coordinates/tests/accuracy/generate_ref_ast.py:254]:
```suggestion
    with Path("data" / fnout).open("wb") as f:
```

File: docs/conftest.py
Comment (nstarman) [reply docs/conftest.py:38]: `pathlib` was introduced in 3.4 and is now pretty feature-complete. Seems a good opportunity to update our approach to paths from the function-oriented manipulation of strings to an object-oriented manipulation of paths. `os.getcwd` isn't deprecated and probably isn't going to be. It's just  the next best thing: superceded.

Comment (neutrinoceros) [reply docs/conftest.py:38]: One thing to keep in mind though is that `Path` objects are much heavier than strings. It doesn't make a measurable difference here but in intensive loops, switching to `pathlib` can add a significant overhead.

File: astropy/config/tests/test_configs.py
Comment (MridulS) [reply astropy/config/tests/test_configs.py:473]: Yes!

File: astropy/convolution/setup_package.py
Comment (MridulS) [reply astropy/convolution/setup_package.py:21]: Not yet! I will send one. I did create one for astropy's extension-helpers https://github.com/astropy/extension-helpers/pull/83

Comment (MridulS) [reply astropy/convolution/setup_package.py:21]: https://github.com/pypa/distutils/pull/230 is up too :)

File: docs/conftest.py
Comment (MridulS) [reply docs/conftest.py:38]: Yeah, and I don't think there are places where we are doing a lot of string/path manipulation inside astropy (maybe I missed something?). A lot of this is indeed aesthetic-y changes. But IMO we should indeed move on to using the obj oriented pathlib way as it make manipulating these path much easier, plus in a typing heavy future of python, `os.PathLike` gives a lot more information than `str`.

**Comment by nstarman** [comment]
@MridulS. This is looking nearly complete. What are the next steps, fixing `get_package_data_` first?

**Comment by nstarman** [comment]
@MridulS, are you interested to finish this PR?
If not, I suggest interested contributors / maintainers cherry-pick parts of this PR on a sub-package basis.

**Comment by MridulS** [comment]
Thanks for the ping! I'll try to finish this up in the next week :)

**Comment by neutrinoceros** [comment]
@MridulS FYI I'm available to revive or take over any lingering PRs of yours if maintainers would like me to.

**Comment by neutrinoceros** [comment]
I can directly push to your branches if you invite me to collaborate on your fork, or I can open brand new PRs, whatever suits you best.

**Comment by pllim** [comment]
Too many conflicts now. Might be best to open a fresh PR at this point. But I will re-open this for now for you. If you do want to continue here, a rebase should reset the timer as far as the bot is concerned.

**Comment by nstarman** [comment]
I'd recommend cherry-picking parts of this PR.

**Comment by astrofrog** [comment]
> I can directly push to your branches if you invite me to collaborate on your fork, or I can open brand new PRs, whatever suits you best.

Just FYI I *think* as a maintainer you can automatically push to PR branches without special permissions

**Comment by neutrinoceros** [comment]
I am no ma...intainer.

https://m.youtube.com/watch?v=W7_c-R7i8F4

**Comment by pllim** [comment]
If you open an alternate PR, please link here and I can close this one (again). Thanks, all!

**Comment by neutrinoceros** [comment]
Yes, I'll probably end up splitting this into smaller, subpackage-targetted PRs. Will try to get to it next week !

**Comment by neutrinoceros** [comment]
I've started chunking this PR in smaller bits starting with #16917, and pushed packaged-targeted patched in #16918, #16918 and #16923 for starters since these bits were already approved here. I'll open all other PRs after #16917 is merged.

**Comment by pllim** [comment]
Since this PR is stale, instead of keeping it open, I have moved the idea to https://github.com/astropy/astropy/issues/16924 . Please continue there. Thanks!

**Pull request closed without merging**
